### PR TITLE
Eip 3475 final patch

### PR DESCRIPTION
**Pull request by dhruvmalik007**
Co-authored : 

- Yu Liu (@yuliu-debond).
- Dhruv Malik (@dhruvmalik007 ).

changes : 
- Added the latest version of the reference implementation.
- Refactored the explanation based on the feedback from the community.
- commenting of functions

### CODE REVIEW

File: EIPS/eip-3475.md
Comment (MicahZoltu) [inline EIPS/eip-3475.md:5]:
```suggestion
author: Yu Liu (@yuliu-debond), Varun Deshpande (@dr-chain), Cedric Ngakam (@drikssy), Dhruv Malik (@GrandGarcon), Samuel Gwlanold Edoumou (@Edoumou), Toufic Batrice (@toufic0710)
```

File: assets/eip-3475/Debond-ERC3475.sol
Comment (MicahZoltu) [inline assets/eip-3475/Debond-ERC3475.sol:20]: Shouldn't be linking to external contracts.  Also, the reference implementation should be as slim as possible, is this a required dependency?

File: assets/eip-3475/Debond-ERC3475.sol
Comment (MicahZoltu) [inline assets/eip-3475/Debond-ERC3475.sol:28]: I'm not sure why you are changing this, `this struct` is more correct I believe.  Plus single space.

**Review by MicahZoltu — COMMENTED** [review]
This is a lot of changes to a final EIP, which are meant to remain pretty stable.

The fact that the reference implementation is receiving a large amount of changes after the EIP is final suggests that it isn't a good reference implementation.  A good reference implementation should be the simplest contract that exemplifies the usage of this standard.  It shouldn't be a full featured application, but rather a nice concise example of how to use this.

EIPs shouldn't be referencing specific products that are on the market today (like MetaMask or Gnosis Safe).  You can refer to them abstractly as something like a "browser wallet" or "contract wallet" if you need to refer to the generalized concept.

File: assets/eip-3475/Debond-ERC3475.sol
Comment (dhruvmalik007) [reply assets/eip-3475/Debond-ERC3475.sol:28]: sure my bad. Actually this was the last check  I did via Grammarly

File: assets/eip-3475/Debond-ERC3475.sol
Comment (dhruvmalik007) [reply assets/eip-3475/Debond-ERC3475.sol]: yes . but eventually we have removed this implementation as this will need considerable interfaces and external contracts. thus to be compliant with eip, the implementation in our assets/ERC3475.sol is sufficient for understanding.

File: assets/eip-3475/Debond-ERC3475.sol
Comment (dhruvmalik007) [reply assets/eip-3475/Debond-ERC3475.sol]: this implementation is removed

**Comment by dhruvmalik007** [comment]
Also @MicahZoltu , we are getting the following validaton error : 

image


So does that mean that we cant refer this repo with the reference, but maybe refer with their common term (NFT's standard) ?

**Comment by dhruvmalik007** [comment]
image


this is the only error remaining. any idea how to resolve this @MicahZoltu?  thanks

**Comment by Pandapip1** [comment]
This is a known @eth-bot bug. We'll review it and @MicahZoltu will manually merge.

File: assets/eip-3475/utils/Context.sol
Comment (Pandapip1) [inline assets/eip-3475/utils/Context.sol:1]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: assets/eip-3475/utils/ownable.sol
Comment (Pandapip1) [inline assets/eip-3475/utils/ownable.sol:1]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: assets/eip-3475/utils/ownable.sol
Comment (Pandapip1) [inline assets/eip-3475/utils/ownable.sol:83]:
```suggestion
}

```

File: assets/eip-3475/utils/Context.sol
Comment (Pandapip1) [inline assets/eip-3475/utils/Context.sol:24]:
```suggestion
}

```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:421]: What was the change here?

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:43]:
```suggestion
pragma solidity ^0.8.0;

```
Please re-add the solidity version, it does help.

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:5]: @Edoumou I imagine this change is legit but I'd like to double-check this.

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:16]: For the line above (L15), I would prefer the following wording change:

```md
This EIP allows the creation of tokenized obligations with abstract on-chain metadata storage. Issuing bonds with multiple redemption data cannot be achieved with existing token standards. This EIP enables each bond class ID to represent a new configurable token type and corresponding to each class, corresponding bond nonces to represent an issuing condition or any other form of data in uint256. Every single nonce of a bond class can have its metadata, supply, and other redemption conditions.
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:394]: What about EIP-1155, another commonly-used token standard?

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:10]:
```suggestion
created: 2021-04-05
requires: 20, 721
```
```suggestion
created: 2021-04-05
requires: 20, 721, 1155
```

Depending on whether you specify compatibility with 1155 or not.

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:19]: Define LP please.

Comment (Pandapip1) [inline EIPS/eip-3475.md:19]:
```suggestion
Current LP tokens are simple [EIP-20](./eip-20.md) tokens with no complex data structure. To allow more complex reward and redemption logic to be stored on-chain, we need a new token standard that:
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:24]:
```suggestion
 - Supports multiple token IDs
 - Can store on-chain metadata
 - Doesn't require a fixed storage pattern
 - Is gas-efficient.
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:26]:
```suggestion
This EIP allows the creation of any obligation with the same interface. It will enable any 3rd party wallet applications or exchanges to read these tokens' balance and redemption conditions. These bonds can also be batched as tradeable instruments. Those instruments can then be divided and exchanged in secondary markets.
```
```suggestion
This proposal allows the creation of any obligation with the same interface. It will enable any 3rd party wallet applications or exchanges to read these tokens' balance and redemption conditions. These bonds can also be batched as tradeable instruments. Those instruments can then be divided and exchanged in secondary markets.
```
`This EIP` is a nit, but `tokens'` is required. Pick your poison.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
First pass

File: EIPS/eip-3475.md
Comment (dhruvmalik007) [reply EIPS/eip-3475.md]: i included your suggestion

File: assets/eip-3475/utils/ownable.sol
Comment (dhruvmalik007) [reply assets/eip-3475/utils/ownable.sol:1]: its from the openzeppelin lib, i can mention that  along with this change

File: EIPS/eip-3475.md
Comment (dhruvmalik007) [reply EIPS/eip-3475.md:5]: sure so for that you need hisverification , @edoumou  can  verify you in DM .

File: EIPS/eip-3475.md
Comment (dhruvmalik007) [reply EIPS/eip-3475.md:394]: yes indeed it can be . the given were example but we can mention

File: EIPS/eip-3475.md
Comment (dhruvmalik007) [reply EIPS/eip-3475.md:10]: ah its weird, in my current branch its already added status: Final

**Comment by dhruvmalik007** [comment]
Hi team, still didnt find out the exact issues facing . specially the CI  bugs of constantStatusError and filePreExistingError. thanks  for the help .

**Comment by MicahZoltu** [comment]
I believe you are encountering a known bug that occurs when you delete an asset associated with an EIP.  If you want to make it go away, you can create a *separate* PR that deletes the asset but doesn't change the status so that the PR is easy to review for editors, then once that is merged this PR can be updated so it doesn't delete a file and we can make sure the bot otherwise passes.

**Comment by dhruvmalik007** [comment]
#5399

**Comment by dhruvmalik007** [comment]
@Pandapip1  @MicahZoltu , any further recommendations before finishing the merge?  thanks

File: assets/eip-3475/utils/ownable.sol
Comment (Pandapip1) [inline assets/eip-3475/utils/ownable.sol:1]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: assets/eip-3475/utils/Context.sol
Comment (Pandapip1) [inline assets/eip-3475/utils/Context.sol:1]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:16]:
```suggestion
This EIP allows the creation of tokenized obligations with abstract on-chain metadata storage. Issuing bonds with multiple redemption data cannot be achieved with existing token standards. This EIP enables each bond class ID to represent a new configurable token type and corresponding to each class, corresponding bond nonces to represent an issuing condition or any other form of data in uint256. Every single nonce of a bond class can have its metadata, supply, and other redemption conditions.
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:18]:
```suggestion
## Motivation

```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:365]:
```suggestion
 This EIP supports batch operations. It allows the user to transfer different bonds along with their metadata to a new address instantaneously in a single transaction. After execution, the new owner holds the right to reclaim the face value of each of the bonds. This mechanism helps with the "packaging" of bonds–helpful in use cases like trades on a secondary market.
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:393]:
```suggestion
**This standard interface is compatible with [EIP-20](./eip-20.md), [EIP-721](./eip-721.md), and [EIP-1155](./eip-1155.md).**
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:401]:
```suggestion
Test cases for the reference implementation are available [here](../assets/eip-3475/ERC3475.test.ts).
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:408]:
```suggestion
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:406]:
```suggestion
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:413]:
```suggestion
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:415]:
```suggestion
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:414]:
```suggestion
- If the owner wants to give a one-time allocation to an address for specific `bonds(classId,bondsId)`, he should call the `function approve()` giving the `Transaction[]` allocated rather than approving all the classes using `setApprovalFor`.
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:416]:
```suggestion
- The `function issue()` can only be called by the issuing contract. The authentication mechanism is unspecified, however, and could conceivably range from fully automated smart contracts to multisigns.
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:387]:
```suggestion
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:388]:
```suggestion
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:389]:
```suggestion
Existing [EIP-20](./eip-20.md) token contracts can issue bonds by delegating the minting role to a bank contract.
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:391]:
```suggestion
```
This should be part of the specification. If this isn't in there, please move it.

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:26]:
```suggestion
```

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:28]: This should be moved to the rationale.

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:30]:
```suggestion
```
If anything, this should be moved to the rationale and mentioned as a trade-off.

File: EIPS/eip-3475.md
Comment (Pandapip1) [inline EIPS/eip-3475.md:32]:
```suggestion
```
Should be in the Abstract, if anything.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Please also respond to my previous feedback

File: assets/eip-3475/utils/Context.sol
Comment (dhruvmalik007) [reply assets/eip-3475/utils/Context.sol:1]: this is not required as these files are now deleted

**Comment by Pandapip1** [comment]
Please fix the eipw errors.

**Review by Edoumou — APPROVED** [review]

**Review by yuliu-debond — APPROVED** [review]

**Comment by dhruvmalik007** [comment]
hi EIP team, 
 Seems that this merge will eventually be blocked because the owner of the branch is now EIP-bot (cc @yuliu-debond ), apart from branch  #5399.

**Comment by yuliu-debond** [comment]
@Pandapip1 @MicahZoltu 
@alita-moore, @mryalamanchi
Hi team,
We are experiencing some difficulties. Could you tell us whats wrong here?
https://github.com/ethereum/EIPs/pull/5399

**Comment by Pandapip1** [comment]
Oh, the reason for the error is due to the deleted files. #5399 must be merged by @SamWilsn, then you can rebase.

**Comment by Pandapip1** [comment]
> Seems that this merge will eventually be blocked because the owner of the branch is now EIP-bot 

This was an intentional change to the @eip-bot workflow. Don't worry, it won't block your PR.

**Comment by dhruvmalik007** [comment]
hi EIP review team, any progress on the side of the EIP bot not being able to merge? thanks.

**Comment by Pandapip1** [comment]
@dhruvmalik007, just checking that you are comfortable with this EIP in its current state. Further modifications after this moves to `final` will be disallowed unless extremely minor.

**Comment by Pandapip1** [comment]
CC also @yuliu-debond @dr-chain @drikssy @Edoumou @toufic0710

**Comment by dhruvmalik007** [comment]
hi @Pandapip1, the last major commits are done and reviewed by me , @drikssy with the implementation . on the side of the deployment you can do that.

**Review by Pandapip1 — APPROVED** [review]

**Comment by xinbenlv** [comment]
Does this ERC want to support EIP-165?

**Comment by dhruvmalik007** [comment]
Thanks @xinbenlv for the question, 

No, we will not be needing ERC-165 as there will not be a standardized implementation by all the different bonds, so there will not be a single signature that every bond holder has to check, and thus there is no need.

**Comment by Pandapip1** [comment]
@eth-bot rerun

**Review by Pandapip1 — APPROVED** [review]

**Comment by dhruvmalik007** [comment]
Hi @Pandapip1 ,  i was checking out the issue  facing the stalling of the auto_review_tests : https://github.com/ethereum/EIPs/blob/master/.github/workflows/auto-review-trigger.yml#L25-L62 .  i didn't  find the issue as the pipeline sees the condition of theapproved and then bot will be doing automatically the tunu of attributing the ERC. along with the PR number.  any troubleshooting on your part ?

**Review by Pandapip1 — APPROVED** [review]

**Comment by Pandapip1** [comment]
Just trying another approval to try to get the bot to merge this.

**Comment by dhruvmalik007** [comment]
@Pandapip1  @SamWilsn  all of the tests are passing but still the blocked version, could you approve the process? thanks

**Comment by yuliu-debond** [comment]
I think otherwise, is best to have a new PR, or do it manually in some way. 

> @Pandapip1 @SamWilsn all of the tests are passing but still the merging is blocked, could you approve again to see? thanks

I think otherwise, is best to have a new PR, or do it manually in some way.

**Review by Pandapip1 — APPROVED** [review]

**Comment by SamWilsn** [comment]
This PR is just waiting on author approval (@yuliu-debond, @dr-chain, @drikssy, @GrandGarcon, @Edoumou, @toufic0710).

**Pull request merged**
